### DISCUSSION TITLE
Reentrancy

### DISCUSSION THREAD

**Discussion by qazwsxedckll** [Tochemey/goakt#320] [Q&A] [answered]
Does goakt support reentrancy？A->A or A->B->C->A. All using synchronous Ask.

```
func (p *Pinger) Receive(ctx goakt.ReceiveContext) {
	switch ctx.Message().(type) {
	case *samplepb.Ping:
		resp := ctx.Ask(ctx.Self(), new(samplepb.Pong))
		ctx.Response(resp)
	case *samplepb.Pong:
		ctx.Response(&samplepb.Pong{
			Reply: "pong",
		})
	default:
		ctx.Unhandled()
	}
}
```

protoactor: A->Future(with response function)->B->C->Future
orleans: Interleaving methods
dapr: bypassing lock if it is in the same request chain

**Comment by Tochemey** [comment]
@qazwsxedckll There is no future concept in Go-Akt. `Ask` expects a response within a time frame.
Also one thing to be aware of the message context is only carried forward when using `Forward`.
I can look into that feature and see whether I would like to support it out of the box or not.
I hope it helps.

**Comment by Tochemey** [comment]
@qazwsxedckll What are the main use cases for such features?

**Reply by qazwsxedckll** [reply]
It is very common that A asks B, B asks C and C need some infomation from A to perform the action. When the chain is long, it is hard to track such loop and sometimes it is inevitable.

**Comment by Tochemey** [comment]
> It is very common that A asks B, B asks C and C need some infomation from A to perform the action. When the chain is long, it is hard to track such loop and sometimes it is inevitable.

This can be easily be done using the pattern `A`--> `Tell`-->`B`-->`Forward`-->`C`-->`Tell`-->`A`. I will read around the use case and see. But I think a `Tell` and `Forward` combination can do the job.

**Reply by qazwsxedckll** [reply]
Someone is waiting for A's response, like a client or a user.

User -> A -> B -> A

Let me clarify. User calls A, waiting for a response. While handling the message, A need some infomation from B, so A calls B waiting a response. However, A doesn't know B also need infomation from itself and B calls A, waiting A to response.

I hope this make sense.

**Comment by Tochemey** [comment]
@qazwsxedckll I have created an issue to look into that https://github.com/Tochemey/goakt/issues/322. Thanks. I hope I have some bandwith to tackle that.

**Comment by Tochemey** [comment, marked as answer]
The implementation is done in this PR https://github.com/Tochemey/goakt/pull/1050
